File: src/piped.py
"""Piped API client with retry/failover, caching, and normalized responses."""
from __future__ import annotations
import time
from urllib.parse import urlencode, urlparse, parse_qs
from .instances import InstanceManager
from .models import Config, PipedError, VideoNotFound

VIDEO_ID_RE = r"^[A-Za-z0-9_-]{11}$"
class PipedClient:
    def __init__(self, config: Config, manager: InstanceManager): self.config, self.manager, self._cache = config, manager, {}
    def playback_url(self, video_id: str, *, autoplay=False, listen=False, quality=None, sponsorblock=None) -> str:
        params = {"v": video_id, "instance": self.manager.select().url}
        if autoplay: params["playerAutoPlay"] = "true"
        if listen: params["listen"] = "true"
        if quality is not None: params["quality"] = str(quality)
        if sponsorblock is not None: params["sponsorblock"] = "true" if sponsorblock else "false"
        return f"{self.config.frontend_url}/watch?{urlencode(params)}"
    async def search(self, query: str, filter="videos", limit=10) -> list[dict]:
        limit = max(1, min(limit, 50)); key = ("search", query, filter, limit); cached = self._get(key, self.config.search_cache_seconds)
        if cached is not None: return cached
        payload, _ = await self._call("/search", {"q": query, "filter": filter}); result = [self._search_item(x) for x in payload.get("items", [])[:limit]]; self._put(key, result); return result
    async def video(self, video_id: str) -> dict:
        key = ("video", video_id); cached = self._get(key, self.config.video_cache_seconds)
        if cached is not None: return cached
        payload, _ = await self._call(f"/streams/{video_id}")
        if not isinstance(payload, dict) or payload.get("title") is None: raise VideoNotFound()
        result = {k: payload.get(k) for k in ("videoId","title","description","uploader","uploaderUrl","duration","views","uploadDate","thumbnailUrl","category","likes","dislikes","livestream","dash","hls","subtitles")}
        result["video_id"] = result.pop("videoId") or video_id; result["channel"] = result.pop("uploader")
        result["streams"] = {"video": self._stream_list(payload.get("videoStreams")), "audio": self._stream_list(payload.get("audioStreams"))}; self._put(key, result); return result
    async def _call(self, path, params=None):
        attempted, last_error = set(), None
        for _ in range(max(1, len(self.manager.instances))):
            state = self.manager.select(attempted); attempted.add(state.url)
            try:
                response = await self.manager._request("GET", state.url.rstrip("/") + path, params=params or {})
                if response.status_code == 404: raise VideoNotFound()
                if response.status_code >= 400: raise RuntimeError(f"HTTP {response.status_code}")
                payload = response.json()
                if not isinstance(payload, dict): raise ValueError("invalid JSON object")
                state.healthy, state.last_success = True, time.time(); return payload, state.url
            except VideoNotFound: raise
            except Exception as exc: last_error = exc; self.manager.mark_failure(state.url)
        raise PipedError(f"Piped backend temporarily unavailable: {last_error}", True)
    @staticmethod
    def _search_item(item):
        kind = item.get("type", "").lower(); raw_url = item.get("url", ""); vid = parse_qs(urlparse(raw_url).query).get("v", [None])[0] or raw_url.rsplit("/", 1)[-1] or item.get("id")
        return {"title": item.get("title"), "channel": item.get("uploaderName") or item.get("channelName"), "duration": item.get("duration"), "views": item.get("views"), "published": item.get("uploadedDate") or item.get("uploadDate"), "thumbnail": item.get("thumbnail") or item.get("thumbnailUrl"), "type": kind, "video_id": vid}
    @staticmethod
    def _stream_list(items): return [{k: item.get(k) for k in ("itag","format","quality","mimeType","codec","width","height","bitrate","contentLength","url","fps","videoOnly") if item.get(k) is not None} for item in (items or [])]
    def _get(self, key, ttl):
        entry = self._cache.get(key); return entry[1] if entry and time.time() - entry[0] < ttl else None
    def _put(self, key, value): self._cache[key] = (time.time(), value)

File: src/piped.js
import { PipedError, VideoNotFound } from './config.js';

export class PipedClient {
  constructor(config, manager) { this.config = config; this.manager = manager; this.cache = new Map(); }
  playbackUrl(videoId, { autoplay = false, listen = false, quality, sponsorblock } = {}) {
    const selected = this.manager.select();
    const params = new URLSearchParams({ v: videoId, instance: selected.url });
    if (autoplay) params.set('playerAutoPlay', 'true'); if (listen) params.set('listen', 'true');
    if (quality !== undefined) params.set('quality', String(quality)); if (sponsorblock !== undefined) params.set('sponsorblock', String(Boolean(sponsorblock)));
    return `${this.config.frontendUrl}/watch?${params}`;
  }
  async search(query, filter = 'videos', limit = 10) {
    limit = Math.max(1, Math.min(50, limit)); const key = `search:${query}:${filter}:${limit}`; const cached = this.get(key, this.config.searchCacheSeconds); if (cached) return cached;
    const { payload } = await this.call('/search', { q: query, filter }); const result = (payload.items ?? []).slice(0, limit).map(normalizeSearchItem); this.put(key, result); return result;
  }
  async video(videoId) {
    const key = `video:${videoId}`; const cached = this.get(key, this.config.videoCacheSeconds); if (cached) return cached;
    const { payload } = await this.call(`/streams/${encodeURIComponent(videoId)}`); if (!payload?.title) throw new VideoNotFound();
    const result = { video_id: payload.videoId ?? videoId, title: payload.title, description: payload.description, channel: payload.uploader, uploaderUrl: payload.uploaderUrl, duration: payload.duration, views: payload.views, published: payload.uploadDate, thumbnail: payload.thumbnailUrl, category: payload.category, likes: payload.likes, dislikes: payload.dislikes, streams: { video: normalizeStreams(payload.videoStreams), audio: normalizeStreams(payload.audioStreams) } };
    this.put(key, result); return result;
  }
  async call(path, params = {}) {
    const attempted = new Set(); let lastError;
    for (let i = 0; i < Math.max(1, this.manager.instances.size); i++) {
      const state = this.manager.select(attempted); attempted.add(state.url);
      try {
        const url = `${state.url.replace(/\/$/, '')}${path}?${new URLSearchParams(params)}`; const response = await this.manager.request(url);
        if (response.status === 404) throw new VideoNotFound(); if (!response.ok) throw new Error(`HTTP ${response.status}`);
        const payload = await response.json(); if (!payload || typeof payload !== 'object') throw new Error('invalid JSON object');
        state.healthy = true; state.lastSuccess = Date.now(); return { payload, instance: state.url };
      } catch (error) { if (error instanceof VideoNotFound) throw error; lastError = error; this.manager.markFailure(state.url); }
    }
    throw new PipedError(`Piped backend temporarily unavailable: ${lastError?.message ?? 'unknown error'}`, true);
  }
  get(key, ttl) { const item = this.cache.get(key); return item && (Date.now() - item.time < ttl * 1000) ? item.value : null; }
  put(key, value) { this.cache.set(key, { time: Date.now(), value }); }
}

function normalizeSearchItem(item) {
  const raw = item.url ?? ''; let videoId = item.id;
  try { videoId = new URL(raw, 'https://piped.local').searchParams.get('v') ?? videoId; } catch {}
  return { title: item.title, video_id: videoId, channel: item.uploaderName ?? item.channelName, duration: item.duration, views: item.views, published: item.uploadedDate ?? item.uploadDate, thumbnail: item.thumbnail ?? item.thumbnailUrl, type: item.type };
}
function normalizeStreams(items = []) { return items.map(item => Object.fromEntries(['itag','format','quality','mimeType','codec','width','height','bitrate','contentLength','url'].filter(k => item[k] !== undefined).map(k => [k, item[k]]))); }
